### PR TITLE
[FLINK-23204] Provide StateBackends access to MailboxExecutor

### PR DESCRIPTION
**Pull request by curcur**
## What is the purpose of the change
This PR provides access to mailbox executor from KeyedStateBackend

There are several places in ChangelogStateBackend that need execute actions from the task thread

DFS writer: collect so far uploaded changes; handle upload results after completion
ChangelogKeyedStateBackend: checkpointing to combine state handles upon upload completion by writer
ChangelogKeyedStateBackend: materialization take a snapshot (sync phase) and handle results of the async phase

Provide access to mailbox executor can simply threading model without a lock.

## Brief changelog
- Move the interface MailboxExecutor to flink-runtime
- Provide mailexecutor access through Environment
- Allow StreamTask to assign mailboxexecutor after initializing it.

## Verifying this change
existing tests

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/execution/Environment.java [hunk 1]
```diff
@@ -37,6 +37,7 @@
 import org.apache.flink.runtime.jobgraph.JobVertexID;
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/execution/Environment.java:37]: nit: can you copy paste PR description into the commit message?

File: flink-runtime/src/main/java/org/apache/flink/runtime/mailbox/FutureTaskWithException.java [hunk 2]
```diff
@@ -16,7 +16,7 @@
  * limitations under the License.
  */
 
-package org.apache.flink.runtime.concurrent;
+package org.apache.flink.runtime.mailbox;
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/mailbox/FutureTaskWithException.java:19]: Why did you move this one? As long as it stays in `flink-runtime` I think it makes more sense to have it in the same package as for example `FutureTask`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/RuntimeEnvironment.java [hunk 3]
```diff
@@ -92,6 +94,10 @@
 
     private final Task containingTask;
 
+    private MailboxExecutor mainMailboxExecutor;
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/RuntimeEnvironment.java:97]: `@Nullable` and ditto for `asyncOperationsThreadPool` an `MockEnvironment`

File: flink-runtime/src/main/java/org/apache/flink/runtime/execution/Environment.java [hunk 4]
```diff
@@ -232,4 +234,20 @@ void acknowledgeCheckpoint(
...
 
     TaskEventDispatcher getTaskEventDispatcher();
+
+    // --------------------------------------------------------------------------------------------
+    //  Fields set in the StreamTask to provide access to mailbox and other runtime resources
+    // --------------------------------------------------------------------------------------------
+
+    default void setMainMailboxExecutor(MailboxExecutor mainMailboxExecutor) {}
+
+    default MailboxExecutor getMainMailboxExecutor() {
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/execution/Environment.java:244]:
```
Optional<MailboxExecutor> getMainMailboxExecutor() { return Optional.empty(); };
```
? and ditto for `getAsyncOperationsThreadPool()`?
Currently this is unchecked `@Nullable`, and `Optional` would solve compile time checks plus solve the problem with `UnsupportedOperationException()`;

Or maybe keep it `MailboxExecutor getMainMailboxExecutor()` but internally inside the implementation do :
```
public  MailboxExecutor getMainMailboxExecutor() {
  checkState(mailboxExecutor != null, "Trying to access not yet initialized mailbox");
  return mailboxExecutor;
}
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/mailbox/FutureTaskWithException.java (hunk 2 above)
Comment (curcur) [reply flink-runtime/src/main/java/org/apache/flink/runtime/mailbox/FutureTaskWithException.java:19]: Yes, you are right. This one does not need to move place as it stays in `flink-runtime`.

It was mistakenly moved when I was trying out different movement places (making all to-be-moved classes in the same package help trying-out.

I moved it back to its original place.

File: flink-runtime/src/main/java/org/apache/flink/runtime/execution/Environment.java (hunk 4 above)
Comment (curcur) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/Environment.java:244]: I do not like Optional, I will add a check in `get`.

Comment (curcur) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/Environment.java:244]: In MockEnvironment, both mailboxExecutor and asycnThreadPool are initialized, and I marked the parameter of `set method` as not nullable. So a `get method` won't return a null;

I will add the check in `RuntimeEnvironment Getter`

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/RuntimeEnvironment.java (hunk 3 above)
Comment (curcur) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/RuntimeEnvironment.java:97]: I mark the `mainMailboxExecutor ` and `asyncOperationsThreadPool` @Nullable in RuntimeEnvironment.

In `MockEnvironment`, both mailboxExecutor and asycnThreadPool are initialized in the constructor, and I marked the parameter of the set method as not nullable. So I did not set `mainMailboxExecutor ` and `asyncOperationsThreadPool` as @Nullable in MockEnvironment, because no way `mainMailboxExecutor ` and `asyncOperationsThreadPool` can be null in `MockEnvironment`.

**Comment by curcur** [comment]
Failure does not seem to be relevant to this fix.

**Comment by curcur** [comment]
Merge it.

**Review by pnowojski — APPROVED** [review]
Thanks for the update @curcur. LGTM assuming green azure.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java [hunk 5]
```diff
@@ -338,6 +338,9 @@ protected StreamTask(
                 Executors.newCachedThreadPool(
                         new ExecutorThreadFactory("AsyncOperations", uncaughtExceptionHandler));
 
+        environment.setMainMailboxExecutor(mainMailboxExecutor);
+        environment.setAsyncOperationsThreadPool(asyncOperationsThreadPool);
```
Comment (rkhachatryan) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:342]: Why isn't IO executor used here?

From its name, `asyncOperationsThreadPool` can be used for anything including "control" actions (like declining checkpoints) or collecting metrics - not only IO.

Though currently it is used for IO only - and that's why we should probably merge the two, likely in favor of Tasks IO executor.

This will also allow to inject it via constructor and make the field immutable.

WDYT?

cc: @pnowojski

Comment (curcur) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:342]: asyncOperationsThreadPool provides async snapshot workers for async checkpoint runnable, which is exactly what we used for materialization in the changelog case.

whether/how to merge IO pool with this async pool is a different topic, which is related but may not be within this ticket's scope.

Comment (rkhachatryan) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:342]: > asyncOperationsThreadPool provides async snapshot workers for async checkpoint runnable, which is exactly what we used for materialization in the changelog case.

I think it's not exactly true - `asyncOperationsThreadPool` only waits for snapshot completion, but upload can happen in other threads (which is the with RocksDB).

Merging these pools is out of scope (if it happens), but should be taken into account IMO.

Comment (curcur) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:342]: > I think it's not exactly true - asyncOperationsThreadPool only waits for snapshot completion, but upload can happen in other threads (which is the with RocksDB).

That's exactly happening here: `asyncOperationsThreadPool` is waiting for a snapshot completion. The difference is that for the normal state backend, async happens during checkpointing; while in materialization(changelog), it happens during periodic materialization.

> Merging these pools is out of scope (if it happens), but should be taken into account IMO.
That part I totally agree, that why `asyncOperationsThreadPool` is reused here.
